src/common/s3.connection.cpp

```diff
@@ -0,0 +1,569 @@
+  std::string& etag)
+{
+    if (bucket_name.empty() || object_name.empty() || 0 == nbytes) {
+        return false;
+    }
+
+    TRACE("Uploading multipart object part %zu for object %s in bucket %s",
+          part_number,
+          object_name.c_str(),
+          bucket_name.c_str());
+
+    std::string_view sv(reinterpret_cast<const char*>(data), nbytes);
+
+    minio::s3::UploadPartArgs args;
+    args.bucket = bucket_name;
+    args.object = object_name;
+    args.part_number = part_number;
+    args.upload_id = upload_id;
+    args.data = sv;
+
+    try {
+        minio::s3::UploadPartResponse response = client_->UploadPart(args);
+        etag = response.etag;
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to upload multipart object part %zu for object %s in "
+             "bucket %s: %s",
+             part_number,
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+bool
+common::S3Connection::complete_multipart_object(
+  const std::string& bucket_name,
+  const std::string& object_name,
+  const std::string& upload_id,
+  const std::list<minio::s3::Part>& parts)
+{
+    if (bucket_name.empty() || object_name.empty()) {
+        return false;
+    }
+
+    TRACE("Completing multipart object %s in bucket %s",
+          object_name.c_str(),
+          bucket_name.c_str());
+    minio::s3::CompleteMultipartUploadArgs args;
+    args.bucket = bucket_name;
+    args.object = object_name;
+    args.upload_id = upload_id;
+    args.parts = parts;
+
+    try {
+        minio::s3::CompleteMultipartUploadResponse response =
+          client_->CompleteMultipartUpload(args);
+
+        return (bool)response;
+    } catch (const std::exception& e) {
+        LOGE("Failed to complete multipart object %s in bucket %s: %s",
+             object_name.c_str(),
+             bucket_name.c_str(),
+             e.what());
+    }
+
+    return false;
+}
+
+common::S3ConnectionPool::S3ConnectionPool(size_t n_connections,
+                                           const std::string& endpoint,
+                                           const std::string& access_key_id,
+                                           const std::string& secret_access_key)
+  : is_accepting_connections_{ true }
+{
+    CHECK(n_connections > 0);
+
+    for (auto i = 0; i < n_connections; ++i) {
+        connections_.push_back(std::make_unique<S3Connection>(
+          endpoint, access_key_id, secret_access_key));
+    }
+}
+
+common::S3ConnectionPool::~S3ConnectionPool() noexcept
+{
+    is_accepting_connections_ = false;
+    cv_.notify_all();
+}
+
+std::unique_ptr<common::S3Connection>
+common::S3ConnectionPool::get_connection()
+{
+    std::unique_lock lock(connections_mutex_);
+    cv_.wait(lock, [this] { return !connections_.empty(); });
+
+    if (should_stop_()) {
+        return nullptr;
+    }
+
+    auto conn = pop_from_connection_pool_();
```

Comment:
I feel like we need to inline `pop_from_connection_pool_` instead of having it be a separate function because it should never be called without acquiring a lock first. Am I missing something?